shared-bindings/audiodelays/Echo.c

```diff
@@ -0,0 +1,300 @@
+// This file is part of the CircuitPython project: https://circuitpython.org
+//
+// SPDX-FileCopyrightText: Copyright (c) 2024 Mark Komus
+//
+// SPDX-License-Identifier: MIT
+
+#include <stdint.h>
+
+#include "shared-bindings/audiodelays/Echo.h"
+#include "shared-module/audiodelays/Echo.h"
+
+#include "shared/runtime/context_manager_helpers.h"
+#include "py/binary.h"
+#include "py/objproperty.h"
+#include "py/runtime.h"
+#include "shared-bindings/util.h"
+
+#define DECAY_DEFAULT 0.7f
+#define MIX_DEFAULT 0.5f
+
+//| class Echo:
+//|     """An Echo effect"""
+//|
+//|     def __init__(
+//|         self,
+//|         max_delay_ms: int = 500,
+//|         delay_ms: BlockInput = 250.0,
+//|         decay: BlockInput = 0.7,
+//|         mix: BlockInput = 0.5,
+//|         buffer_size: int = 512,
+//|         sample_rate: int = 8000,
+//|         bits_per_sample: int = 16,
+//|         samples_signed: bool = True,
+//|         channel_count: int = 1,
+//|     ) -> None:
+//|         """Create a Echo effect that echos an audio sample every set number of milliseconds.
+//|
+//|         :param int max_delay_ms: The maximum delay the echo can be
+//|         :param BlockInput delay_ms: The current echo delay
+//|         :param BlockInput decay: The rate the echo fades. 0.0 = instant; 1.0 = never.
```

Comment:
Would 1.0 require a buffer of infinite size? How should you handle a case where the buffer is too small? Maybe mention how `max_delay_ms` impacts the internal echo buffer.

Reply:
A `decay` of 1.0 will cause an infinite repeat of the audio signal. It shouldn't have any effect on buffer size. I think it's better to allow for this value rather than prevent it. There are some situation where infinite repeat is desired (imagine a momentary "infinite" button). Some analog delay effects actually allow for a feedback/decay value greater than 1.0 to cause self-oscillation. That's something we want to avoid here, but just a bit of context.